### PR TITLE
Collection of adapter fixes

### PR DESCRIPTION
**Pull request by Joshua-Ashton**

### CODE REVIEW

**Comment by DadSchoorse** [comment]
Calling `vkGetPhysicalDeviceProperties2` is not allowed on 1.0 devices. You should check with `vkGetPhysicalDeviceProperties` if a device supports `VK_API_VERSION_1_1`.

**Comment by Joshua-Ashton** [comment]
We don't ever make a 1.0 instance or device.

**Comment by DadSchoorse** [comment]
That doesn't matter, if you create a 1.1 device and call `vkEnumeratePhysicalDevices`, you will get devices which only support 1.0 as well.

**Comment by doitsujin** [comment]
> We don't ever make a 1.0 instance or logical device, can the physical device still expose that? That's annoying.

Yes, and this is exactly why things like the device select layer and old DXVK versions break when there's a 1.0 driver present.

**Comment by Joshua-Ashton** [comment]
Okay, will fix that. Mildly annoying.

**Comment by Joshua-Ashton** [comment]
Hopefully resolved.

File: libs/d3d12/main.c [hunk 1]
```diff
@@ -233,39 +197,46 @@ static VkPhysicalDevice d3d12_get_vk_physical_device(struct vkd3d_instance *inst
...
+        properties2.pNext = &id_properties;
 
-            if (!memcmp(id_properties.deviceLUID, &adapter_desc->AdapterLuid, VK_LUID_SIZE))
-            {
-                vk_physical_device = vk_physical_devices[i];
-                break;
-            }
+        pfn_vkGetPhysicalDeviceProperties2(vk_physical_devices[i], &properties2);
+
+        if (!memcmp(id_properties.deviceLUID, &adapter_desc->AdapterLuid, VK_LUID_SIZE))
```
Comment (DadSchoorse) [inline libs/d3d12/main.c:219]: You should check `id_properties.deviceLUIDValid` first.

File: libs/d3d12/main.c [hunk 2]
```diff
@@ -233,39 +197,46 @@ static VkPhysicalDevice d3d12_get_vk_physical_device(struct vkd3d_instance *inst
...
-            memset(&id_properties, 0, sizeof(id_properties));
-            id_properties.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ID_PROPERTIES;
+        /* Skip over < 1.1 physical devices because... 1.1 instances
+         * still expose those apparently... */
+        if (properties2.properties.apiVersion < VK_API_VERSION_1_1)
+            continue;
 
-            properties2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2;
-            properties2.pNext = &id_properties;
+        memset(&id_properties, 0, sizeof(id_properties));
```
Comment (DadSchoorse) [inline libs/d3d12/main.c:211]: Why `memset` the whole struct instead of just setting `pNext` to `NULL`?

**Review by DadSchoorse — CHANGES_REQUESTED** [review]
comments

File: libs/d3d12/main.c (hunk 2 above)
Comment (Joshua-Ashton) [reply libs/d3d12/main.c:211]: The existing code did this, but, fixed.

File: libs/d3d12/main.c (hunk 1 above)
Comment (Joshua-Ashton) [reply libs/d3d12/main.c:219]: Resolved.

**Review by DadSchoorse — APPROVED** [review]

File: libs/d3d12/main.c [hunk 3]
```diff
@@ -233,39 +197,46 @@ static VkPhysicalDevice d3d12_get_vk_physical_device(struct vkd3d_instance *inst
...
-        TRACE("Matching adapters by LUIDs.\n");
+        pfn_vkGetPhysicalDeviceProperties(vk_physical_devices[i], &properties2.properties);
 
-        for (i = 0; i < count; ++i)
-        {
-            memset(&id_properties, 0, sizeof(id_properties));
-            id_properties.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ID_PROPERTIES;
+        /* Skip over < 1.1 physical devices because... 1.1 instances
+         * still expose those apparently... */
+        if (properties2.properties.apiVersion < VK_API_VERSION_1_1)
```
Comment (doitsujin) [inline libs/d3d12/main.c:208]: Should probably check `VKD3D_MIN_API_VERSION` (which is currently 1_1 but may be bumped to 1_2 at some point) and remove the comment above, instead WARN that we skipped an adapter.

**Comment by doitsujin** [comment]
I think there's more that can be done here, since the `vkd3d_select_physical_device` cruft still exists. That should probably be removed and we should make sure that we always find a valid non-null physical device that supports all required extensions right away, and if we can't, error out.

**Review by doitsujin — CHANGES_REQUESTED** [review]

**Comment by Joshua-Ashton** [comment]
`vkd3d_select_physical_device` is used for native builds. We can clean that up later, I want to remove the whole vkd3d-utils thing and consolidate it into d3d12 eventually.

**Review by HansKristian-Work — COMMENTED** [review]
Don't have further concerns than existing comments.

File: libs/d3d12/main.c (hunk 3 above)
Comment (Joshua-Ashton) [reply libs/d3d12/main.c]: Resolved.

**Review by doitsujin — APPROVED** [review]

**Pull request merged**